lib/runtime-core/src/typed_func.rs

```diff
@@ -570,84 +707,143 @@ macro_rules! impl_traits {
+                    let self_pointer = wrap::<Error, FN> as *const vm::Func;
+
+                    // Get the real host function to call.
+                    let (func, func_index): (&FN, ImportedFuncIndex) = wrap_get_func(vmctx.import_backing, self_pointer);
+
+                    // Read the signature.
+                    let module_info = &unsafe{ &*vmctx.module }.info;
+                    let func_signature_index = module_info.func_assoc[func_index.convert_up(&module_info)];
+                    let func_signature = &module_info.signatures[func_signature_index];
+
+                    // Store all arguments in a single array.
+                    let arguments = &[ $( $x ),* ];
 
                     // Catch unwind in case of errors.
-                    let err = match panic::catch_unwind(
-                        panic::AssertUnwindSafe(
-                            || {
-                                func($( WasmExternType::from_native($x), )* ).report()
+                    let err = match panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                        // Map `u64` to `Value` based on the signature.
+                        let inputs: Vec<Value> = arguments
+                            .iter()
+                            .zip(func_signature.params().iter())
+                            .map(
+                                |(argument, ty)| {
+                                    match ty {
+                                        Type::I32 => i32::from_binary(*argument).into(),
+                                        Type::I64 => i64::from_binary(*argument).into(),
+                                        Type::F32 => f32::from_binary(*argument).into(),
+                                        Type::F64 => f64::from_binary(*argument).into(),
+                                        _ => unimplemented!("Variadic host function doesn't support `v128` values."),
+                                    }
+                                }
+                            )
+                            .collect();
+
+                        let vmctx = unsafe { &mut *(vmctx as *const _ as *mut _) };
+
+                        func(vmctx, inputs.as_slice())
+                        //   ^^^^^ The imported function
+                        //         expects `vm::Ctx` as first
+                        //         argument; provide it.
+                    })) {
+                        Ok(Ok(returns)) => return {
+                            let mut bytes = [0u8; count_idents!( $( $x ),* ) * mem::size_of::<u64>()];
+                            let mut index = 0;
+
+                            for value in returns.iter() {
+                                match value {
+                                    Value::I32(value) => {
+                                        let next_index = index + mem::size_of::<i32>();
+
+                                        NativeEndian::write_i32(&mut bytes[index..next_index], *value);
```

Comment:
This appears to be tightly packing, so `let mut bytes = [0u8; count_idents!( $( $x ),* ) * mem::size_of::<u64>()];` will have extra space at the end if you have an 32bit values; is that correct for alignment?

One other thing that could be important when doing unsafe stuff: `[u8; N]` has alignment 1.  Therefore it's undefined behavior in general to use that array of bytes as a type of higher alignment in Rust.  If you want a higher alignment, we need to construct the array as a type of something like `[u64; N / 8]`

Reply:
>  will have extra space at the end if you have an 32bit values

It is correct yes. Since we already know the size of the values at runtime, I decided to go that way: Allocating enough space for the largest values. The resulting value needs to be allocated on the stack.

> is that correct for alignment?

I believe it is yes, since the `S0`, `S1`, `S2` & co. structs have a C representation, I think… well, no, C hasn't a packing alignement, it has a padding alignment. I should address that!